execution/block-partitioner/src/lib.rs

```diff
@@ -3,34 +3,244 @@
+/// - Required edge set matches dependency edge set.
+/// - Before the last round, there is no in-round cross-shard dependency.
+///
+/// Also print a summary of the partitioning result.
+pub fn verify_partitioner_output(
+    before_partition: &Vec<AnalyzedTransaction>,
+    after_partition: &Vec<SubBlocksForShard<AnalyzedTransaction>>,
+) {
+    let old_txn_id_by_txn_hash: HashMap<HashValue, usize> = HashMap::from_iter(
+        before_partition
+            .iter()
+            .enumerate()
+            .map(|(tid, txn)| (txn.test_only_hash(), tid)),
+    );
+
+    let mut total_comm_cost = 0;
+    let num_txns = before_partition.len();
+    let num_shards = after_partition.len();
+    let num_rounds = after_partition
+        .first()
+        .map(|sbs| sbs.sub_blocks.len())
+        .unwrap_or(0);
+    for sub_block_list in after_partition.iter().take(num_shards).skip(1) {
+        assert_eq!(num_rounds, sub_block_list.sub_blocks.len());
+    }
+    let mut old_tids_by_sender: HashMap<Sender, Vec<usize>> = HashMap::new();
+    let mut old_tids_seen: HashSet<usize> = HashSet::new();
+    let mut edge_set_from_src_view: HashSet<(usize, usize, usize, HashValue, usize, usize, usize)> =
+        HashSet::new();
+    let mut edge_set_from_dst_view: HashSet<(usize, usize, usize, HashValue, usize, usize, usize)> =
+        HashSet::new();
+    for round_id in 0..num_rounds {
+        for (shard_id, sub_block_list) in after_partition.iter().enumerate() {
+            let sub_block = sub_block_list.get_sub_block(round_id).unwrap();
+            let mut cur_sub_block_inbound_costs_by_key_src_pair: HashMap<
+                (RoundId, ShardId, StateKey),
+                u64,
+            > = HashMap::new();
+            let mut cur_sub_block_connectivity_by_key_dst_pair: HashMap<
+                (RoundId, ShardId, StateKey),
+                u64,
+            > = HashMap::new();
+            for (local_tid, td) in sub_block.transactions.iter().enumerate() {
+                let sender = td.txn.sender();
+                let old_tid = *old_txn_id_by_txn_hash
+                    .get(&td.txn().test_only_hash())
+                    .unwrap();
+                old_tids_seen.insert(old_tid);
+                old_tids_by_sender
+                    .entry(sender)
+                    .or_insert_with(Vec::new)
+                    .push(old_tid);
+                let tid = sub_block.start_index + local_tid;
+                for loc in td.txn.write_hints().iter() {
+                    let key = loc.clone().into_state_key();
+                    let key_str = CryptoHash::hash(&key).to_hex();
+                    println!(
+                        "MATRIX_REPORT - round={}, shard={}, old_tid={}, new_tid={}, write_hint={}",
+                        round_id, shard_id, old_tid, tid, key_str
+                    );
+                }
+                for (src_tid, locs) in td.cross_shard_dependencies.required_edges().iter() {
```

Comment:
Do not yet have a good way to break down - I tried pulling them out and the whole file gets even longer because several objects with long types are passed down across multiple layers...